### PR TITLE
feat(product analytics): Remove insight exit pop up

### PR DESCRIPTION
**Pull request by k11kirky**
## Problem

When exploring with an new insight or editing an existing insight users are frustrated with the web alert pop up to confirm leaving the page. Also to share an explore you first have to save it and then share the url.

## Changes

Removed popup from temporary insights or insights that are being edited.
Instead saving the query to both the url and local storage

Allow user to return to temp insights via a new button and via url

Screenshot 2024-12-12 at 4 48 00 PM

Example url
http://localhost:8000/project/1/insights/new#q=%7B%22kind%22%3A%22InsightVizNode%22%2C%22source%22%3A%7B%22kind%22%3A%22TrendsQuery%22%2C%22series%22%3A%5B%7B%22kind%22%3A%22EventsNode%22%2C%22event%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22math%22%3A%22first_time_for_user%22%7D%5D%2C%22trendsFilter%22%3A%7B%7D%7D%7D%20

## How did you test this code

Todo

### CODE REVIEW

File: frontend/src/scenes/insights/utils.tsx [hunk 1]
```diff
@@ -433,3 +434,9 @@ export function insightUrlForEvent(event: Pick<EventType, 'event' | 'properties'
 
     return query ? urls.insightNew(undefined, undefined, query) : undefined
 }
+
+export function isQueryTooLarge(query: Node<Record<string, any>>): boolean {
+    // Chrome has a 2MB limit for the HASH params, limit ours at 1MB
+    const queryLength = JSON.stringify(query).length
+    return queryLength > 1024 * 1024
```
Comment (anirudhpillai) [inline frontend/src/scenes/insights/utils.tsx:441]: might be wrong here but you might need something [more complex](https://stackoverflow.com/a/68791382/5540417) to check number of bytes used by a JS string? :)

https://stackoverflow.com/a/52254083/5540417

File: frontend/src/scenes/saved-insights/SavedInsights.tsx [hunk 2]
```diff
@@ -671,6 +683,13 @@ export function SavedInsights(): JSX.Element {
                         <SavedInsightsEmptyState />
                     ) : (
                         <>
+                            {draftQuery?.query && (
+                                <div className="text-muted-alt mb-4">
```
Comment (anirudhpillai) [inline frontend/src/scenes/saved-insights/SavedInsights.tsx:687]: Might make sense to show this on the new insight page as well?
when someone tries to create a new insight again (clicking the + button without going to the saved insights page) they might want to go back to the previous state they had

File: frontend/src/scenes/insights/insightDataLogic.tsx [hunk 3]
```diff
@@ -187,21 +193,62 @@ export const insightDataLogic = kea<insightDataLogicType>([
                 actions.setInsightData({ ...values.insightData, result })
             }
         },
-        loadInsightSuccess: ({ insight }) => {
```
Comment (anirudhpillai) [inline frontend/src/scenes/insights/insightDataLogic.tsx:190]: bit confused why this was here in the first place 😅 , maybe @thmsobrmlr has more context.
Though wanted to check if you had any reasons for removing it? If it's just not needed can also remove it from line 58
```
insightLogic,
['setInsight', 'loadInsightSuccess'],
```

**Review by anirudhpillai — COMMENTED** [review]
Great work on this! Such a massive UX improvement for users by removing the popup. Love how we can get back to the unsaved insight. Just one thing I noticed is that when you're adding a new insight from a dashboard (or when you click back on the + sign) you are brought back to the unsaved insight.


https://github.com/user-attachments/assets/cd01cf8c-ecbd-4549-9abf-cbb59980da1b



Think it might be better to give the user an option to revert to the unsaved state instead of starting off from it? Especially in case of adding insights on dashboards where the unsaved insight might have been quite different + if unsaved insight had a lot of filters, the user will need to remove everything to get the blank slate to start from. I might be wrong here, so will let others chime in as well :)

Looking great otherwise!

**Comment by aspicer** [comment]
A thought about users with multiple teams (soon to be called environments): for users with multiple teams, we would want to scope these localstorage queries to only show up in a given team

File: frontend/src/scenes/saved-insights/ReloadInsight.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        }
+    }
+
+    if (!draftQuery?.query) {
+        return <> </>
+    }
+    return (
+        <div className="text-muted-alt mb-4">
+            You have an unsaved insight from {new Date(draftQuery.timestamp).toLocaleString()}.{' '}
+            <Link to={urls.insightNew(undefined, null, draftQuery.query)}>Click here</Link> to view it.
```
Comment (aspicer) [inline frontend/src/scenes/saved-insights/ReloadInsight.tsx:28]: small nit: how do you feel about changing this to say "click here to load it" if you are already on the insight editor?

**Comment by aspicer** [comment]
I tried this out and it seems great.

I did struggle with loading in the mental model of how the unsaved insight interacts with edit mode.

If you have an unsaved insight, and then browse to any insight and click edit, and then make any changes there, it overrides your unsaved insight.

If you then go and "view" your unsaved insight, it doesn't take you back to edit mode for the saved insight, but takes you to a generic edit mode for a new insight. If you edit it and save, you then get a modified version of the insight you went to edit.

I could see this being confusing. I don't have a slam dunk answer here to make it simple, but potentially we could make the "unsaved insight" feature only applies to new insights and we *always* show the save modal if you are in "edit" mode for an existing insight and try to navigate away.

I could see a more complex interaction with editing saved insights in the future, but I don't want to scope creep this too much.

If you have a better model for how this interaction could work, open to hearing it.

**Review by aspicer — APPROVED** [review]
This looks great!

**Pull request merged**